### PR TITLE
GPULink support

### PR DESCRIPTION
**Pull request by wjp**
This adds an interface for externally managed (and potentially persistent) GPU memory.

### CODE REVIEW

File: include/astra/Float32Data3DGPU.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
+/*
+-----------------------------------------------------------------------
+Copyright: 2010-2015, iMinds-Vision Lab, University of Antwerp
+           2014-2015, CWI, Amsterdam
```
Comment (kohr-h) [inline include/astra/Float32Data3DGPU.h]: These sections could use an update I guess?

File: include/astra/Float32Data3DGPU.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+Contact: astra@uantwerpen.be
+Website: http://sf.net/projects/astra-toolbox
```
Comment (kohr-h) [inline include/astra/Float32Data3DGPU.h]: Link new homepage?

File: include/astra/Float32Data3DGPU.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+#include "../../cuda/3d/mem3d.h"
+
+namespace astra {
+
+
+astraCUDA3d::MemHandle3D wrapHandle(float *D_ptr, unsigned int x, unsigned int y, unsigned int z, unsigned int pitch);
+
+
+/** 
+ * This class represents a three-dimensional block of float32ing point data.
```
Comment (kohr-h) [inline include/astra/Float32Data3DGPU.h]: float32ing?

File: include/astra/Float32Data3DGPU.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	void _unInit();
+
+	/** Initialization. Initializes an instance of the CFloat32Data3DGPU class.
+	 * Can only be called by derived classes.
+	 *
+	 * This function does not set m_bInitialized to true if everything is ok.
+	 *
+	 * @param _iWidth width of the 2D data (x-axis), must be > 0
+	 * @param _iHeight height of the 2D data (y-axis), must be > 0 
+	 * @param _iDepth depth of the 2D data (z-axis), must be > 0
```
Comment (kohr-h) [inline include/astra/Float32Data3DGPU.h]: 2D?

File: python/astra/data3d.py [hunk 5]
```diff
@@ -52,11 +54,11 @@ def link(datatype, geometry, data):
     :returns: :class:`int` -- the ID of the constructed object.
     
     """
-    if not isinstance(data,np.ndarray):
+    if not isinstance(data,np.ndarray) and not isinstance(data,GPULink):
         raise ValueError("Input should be a numpy array")
```
Comment (kohr-h) [inline python/astra/data3d.py]: change this, too?

File: python/astra/data3d.py [hunk 6]
```diff
@@ -52,11 +54,11 @@ def link(datatype, geometry, data):
...
-    if not data.dtype==np.float32:
+    if not isinstance(data,GPULink) and not data.dtype==np.float32:
```
Comment (kohr-h) [inline python/astra/data3d.py]: `and data.dtype != np.float32`
The logic here (and below) is that anything which is not a `GPULink` is a `numpy.ndarray`, which is true *currently*. Later changes will be easier if this code tests for `isinstance(data, np.ndarray)` explicitly.

File: python/astra/data3d_c.pyx [hunk 7]
```diff
@@ -65,12 +70,19 @@ def create(datatype,geometry,data=None, link=False):
...
-        raise Exception("The dimensions of the data do not match those specified in the geometry.")
+    cdef CFloat32CustomMemory * pCustom = NULL
+    IF HAVE_CUDA==True:
+        cdef MemHandle3D hnd
+
+    if link:
+        if isinstance(data, GPULink):
+            s = geom_size(geometry)
+            if geom_size(geometry) != ( data.z, data.y, data.x ):
+                raise Exception("The dimensions of the data do not match those specified in the geometry.")
```
Comment (kohr-h) [inline python/astra/data3d_c.pyx]: While changing this, the error type and message should be improved as well, to minimize user frustration.
In particular, print received and expected values.
- `Exception` is far too general, raise `ValueError` instead
- Error message could be something like
  ```
  'data shape does not match sizes as specified in the '
  'geometry: {} != {}'.format((data.z, data.y, data.x), s)
  ```
- You define `s` but never use it :-)

File: python/astra/pythonutils.py [hunk 8]
```diff
@@ -61,3 +61,21 @@ def geom_size(geom, dim=None):
         s = s[dim]
 
     return s
+
+class GPULink:
```
Comment (kohr-h) [inline python/astra/pythonutils.py]: For Python 2, you want to inherit from `object` to get new-style classes.

File: samples/python/s021_pygpu.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+angles = np.linspace(0, 2 * np.pi, 180, False)
+proj_geom = astra.create_proj_geom('cone', 1.0, 1.0, 128, 192, angles, 1000, 0)
+
+# Create a simple hollow cube phantom
+cube = np.zeros((128, 128, 128))
+cube[17:113, 17:113, 17:113] = 1
+cube[33:97, 33:97, 33:97] = 0
+
+# Create a pygpu gpuarray for the phantom
+vol_gpuarr = pygpu.gpuarray.array(cube, dtype='float32', order='C')
```
Comment (kohr-h) [inline samples/python/s021_pygpu.py]: It should actually work to skip the numpy array and to the exact steps as above for `cube` on the GPU array. They also support this Numpy style slicing syntax.

**Comment by wjp** [comment]
Thanks! These commits should address your comments.

**Comment by kohr-h** [comment]
Sweet!

**Pull request merged**